amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,179 @@
+        const amrex::Real y = prob_lo[1] + (j + 0.5) * dx[1];
+        const amrex::Real z = prob_lo[2] + (k + 0.5) * dx[2];
+        amrex::Real xstart_damping = 0;
+        amrex::Real ystart_damping = 0;
+        amrex::Real xend_damping = 0;
+        amrex::Real yend_damping = 0;
+        amrex::Real xi_end = (x - start_east) / (prob_hi[0] - start_east);
+        amrex::Real xi_start = (start_west - x) / (start_west - prob_lo[0]);
+        xi_start = sponge_west * std::max(xi_start, 0.0);
+        xi_end = sponge_east * std::max(xi_end, 0.0);
+        xstart_damping = sponge_west * sponge_strength * xi_start * xi_start;
+        xend_damping = sponge_east * sponge_strength * xi_end * xi_end;
+        amrex::Real yi_end = (y - start_north) / (prob_hi[1] - start_north);
+        amrex::Real yi_start = (start_south - y) / (start_south - prob_lo[1]);
+        yi_start = sponge_south * std::max(yi_start, 0.0);
+        yi_end = sponge_north * std::max(yi_end, 0.0);
+        ystart_damping = sponge_strength * yi_start * yi_start;
+        yend_damping = sponge_strength * yi_end * yi_end;
+        amrex::Real spongeVelX = 0.0;
+        amrex::Real spongeVelY = 0.0;
+        amrex::Real spongeVelZ = 0.0;
+        amrex::Real residual = 1000;
+        amrex::Real height_error = 0.0;
+        for (unsigned ii = 0; ii < vsize; ++ii) {
+            height_error = std::abs(z - device_vel_ht[ii]);
+            if (height_error < residual) {
+                residual = height_error;
+                const unsigned ix = 3 * ii;
+                const unsigned iy = 3 * ii + 1;
+                const unsigned iz = 3 * ii + 2;
+                spongeVelX = device_vel_vals[ix];
+                spongeVelY = device_vel_vals[iy];
+                spongeVelZ = device_vel_vals[iz];
+            }
+        }
+        amrex::Real Dxz = 0.0;
+        amrex::Real Dyz = 0.0;
+        amrex::Real bcforcing_x = 0;
+        amrex::Real bcforcing_y = 0;
+        const amrex::Real ux1 = vel(i, j, k, 0);
+        const amrex::Real uy1 = vel(i, j, k, 1);
+        const amrex::Real uz1 = vel(i, j, k, 2);
+        const amrex::Real m = std::sqrt(ux1 * ux1 + uy1 * uy1 + uz1 * uz1);
+        if (drag(i, j, k) == 1 && (!is_laminar)) {
+            const amrex::Real ux2 = vel(i, j, k + 1, 0);
+            const amrex::Real uy2 = vel(i, j, k + 1, 1);
+            const amrex::Real m2 = std::sqrt(ux2 * ux2 + uy2 * uy2);
+            const amrex::Real kappa = 0.41;
+            const amrex::Real z0 = std::max(terrainz0(i, j, k), 1e-4);
+            const amrex::Real ustar = m2 * kappa / std::log(1.5 * dx[2] / z0);
+            const amrex::Real uTarget =
+                ustar / kappa * std::log(0.5 * dx[2] / z0);
+            const amrex::Real uxTarget =
+                uTarget * ux2 / (1e-5 + std::sqrt(ux2 * ux2 + uy2 * uy2));
+            const amrex::Real uyTarget =
+                uTarget * uy2 / (1e-5 + std::sqrt(ux2 * ux2 + uy2 * uy2));
```

Comment:
these 1e-5 numbers seem kind of large to just avoid dividing by 0, which seems like what's happening here.

Reply:
Yes it was added to avoid divide by zero. I can reduce it. Is there any default amex variable I can use?

Reply:
@marchdf could you comment on a better way to do this?

Reply:
I would think the best practice is to set
`const auto tiny = std::numeric_limits<amrex::Real>::epsilon()`
before the loop and then use tiny throughout.